vectorstores/pgvector/pgvector.go

```diff
@@ -0,0 +1,329 @@
+	tx, err := s.conn.Begin(ctx)
+	if err != nil {
+		return err
+	}
+	// inspired by
+	// https://github.com/langchain-ai/langchain/blob/v0.0.340/libs/langchain/langchain/vectorstores/pgvector.py#L167
+	// The advisor lock fixes issue arising from concurrent
+	// creation of the vector extension.
+	// https://github.com/langchain-ai/langchain/issues/12933
+	// For more information see:
+	// https://www.postgresql.org/docs/16/explicit-locking.html#ADVISORY-LOCKS
+	if _, err := tx.Exec(ctx, "SELECT pg_advisory_xact_lock(1573678846307946495)"); err != nil {
+		return err
+	}
+	if _, err := tx.Exec(ctx, "CREATE EXTENSION IF NOT EXISTS vector"); err != nil {
+		return err
+	}
+	return tx.Commit(ctx)
+}
+
+func (s Store) createCollectionTableIfNotExists(ctx context.Context) error {
+	tx, err := s.conn.Begin(ctx)
+	if err != nil {
+		return err
+	}
+	// inspired by
+	// https://github.com/langchain-ai/langchain/blob/v0.0.340/libs/langchain/langchain/vectorstores/pgvector.py#L167
+	// The advisor lock fixes issue arising from concurrent
+	// creation of the vector extension.
+	// https://github.com/langchain-ai/langchain/issues/12933
+	// For more information see:
+	// https://www.postgresql.org/docs/16/explicit-locking.html#ADVISORY-LOCKS
+	if _, err = tx.Exec(ctx, "SELECT pg_advisory_xact_lock(1573678846307946494)"); err != nil {
+		return err
+	}
+	sql := fmt.Sprintf(`CREATE TABLE IF NOT EXISTS %s (
+	name varchar,
+	cmetadata json,
+	"uuid" uuid NOT NULL,
+	PRIMARY KEY (uuid))`, s.collectionTableName)
+	if _, err = tx.Exec(ctx, sql); err != nil {
+		return err
+	}
+	return tx.Commit(ctx)
+}
+
+func (s Store) createEmbeddingTableIfNotExists(ctx context.Context) error {
+	tx, err := s.conn.Begin(ctx)
+	if err != nil {
+		return err
+	}
+	// inspired by
+	// https://github.com/langchain-ai/langchain/blob/v0.0.340/libs/langchain/langchain/vectorstores/pgvector.py#L167
+	// The advisor lock fixes issue arising from concurrent
+	// creation of the vector extension.
+	// https://github.com/langchain-ai/langchain/issues/12933
+	// For more information see:
+	// https://www.postgresql.org/docs/16/explicit-locking.html#ADVISORY-LOCKS
+	if _, err := tx.Exec(ctx, "SELECT pg_advisory_xact_lock(1573678846307946493)"); err != nil {
+		return err
+	}
+	sql := fmt.Sprintf(`CREATE TABLE IF NOT EXISTS %s (
+	collection_id uuid,
+	embedding vector,
+	document varchar,
+	cmetadata json,
+	custom_id varchar,
+	"uuid" uuid NOT NULL,
+	CONSTRAINT langchain_pg_embedding_collection_id_fkey 
+	FOREIGN KEY (collection_id) REFERENCES %s (uuid) ON DELETE CASCADE,
+PRIMARY KEY (uuid))`, s.embeddingTableName, s.collectionTableName)
+	if _, err = tx.Exec(ctx, sql); err != nil {
+		return err
+	}
+	return tx.Commit(ctx)
+}
+
+func (s Store) AddDocuments(ctx context.Context, docs []schema.Document, options ...vectorstores.Option) error {
+	opts := s.getOptions(options...)
+	if opts.Embedder != nil || opts.ScoreThreshold != 0 || opts.Filters != nil || opts.NameSpace != "" {
```

Comment:
This is confusing to read, it's unhappy if `Embedder` is provided? Can you explain in a comment what's going on?

Reply:
I previously thought that using the embedder already in Store would be sufficient, but it makes sense to use the embedder override in opts.

Update code in commit 1faf360fc0a5ec31fcdb2560396c1680d5216fbe